File: crates/api/src/commands.rs
//! Writes. Every one moves something, and each says what it moves it to.
//!
//! **Split from the route table rather than from a line count.** What separates
//! these from `queries` is not size: a command decodes a body, may answer 201,
//! and has refusals that mean the machine would not admit the move. A caller
//! reading one of these needs to know what the Job looks like afterwards, which
//! is what each doc below states and no query has to.
//!
//! The 400 for a body that would not parse is the transport's own and never the
//! daemon's — see `crate::answers::undecodable`. Nothing downstream was asked.

use axum::body::Bytes;
use axum::extract::State;
use axum::http::StatusCode;
use axum::response::Response;
use ipc::{
    CapRaise, ChangesRequested, ChosenAnswer, FileReport, JobRequest, Overruled, ProposeJob,
    Redirection, RemarksTakenUp, RestartRequested, StopProposal, TurnRaise,
};

use crate::answers::{answer, refused, undecodable};
use crate::daemon::Commands;
use crate::reference::Resolved;
use crate::routes::Served;

pub(crate) async fn propose_job<D: Commands>(
    State(served): State<Served<D>>,
    body: Bytes,
) -> Response {
    let proposal: ProposeJob = match ipc::decode("proposed Job", &body) {
        Ok(proposal) => proposal,
        // 400 is the transport's own refusal and never the daemon's — the
        // bytes did not become a request, so nothing downstream was asked.
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().propose_job(proposal).await {
        // 201: the Job now exists, at the approval gate. It is not running, and
        // nothing here approves it.
        Ok(job) => answer(StatusCode::CREATED, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The other way a Job reaches the gate: a person describes the work and the
/// proposer reads it. **The same 201 and the same gate** — what differs is who
/// filled the workflow in, and that one request can be several Jobs.
pub(crate) async fn propose_from_request<D: Commands>(
    State(served): State<Served<D>>,
    body: Bytes,
) -> Response {
    let request: JobRequest = match ipc::decode("request", &body) {
        Ok(request) => request,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().propose_from_request(request).await {
        Ok(job) => answer(StatusCode::CREATED, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Stop a proposal that is still out.
///
/// **200 on both arms**, including the one where there was nothing to stop.
/// `ipc::ProposalStopped` carries which, and a 404 for a proposal that has just
/// landed would report a failure to somebody whose Jobs are on the board.
pub(crate) async fn stop_proposal<D: Commands>(
    State(served): State<Served<D>>,
    body: Bytes,
) -> Response {
    let stopping: StopProposal = match ipc::decode("a proposal to stop", &body) {
        Ok(stopping) => stopping,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().stop_proposal(stopping.proposal_id).await {
        Ok(stopped) => answer(StatusCode::OK, &stopped, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

pub(crate) async fn approve_dispatch<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().approve_dispatch(job.id()).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The person takes the work. **The counterpart to `approve_dispatch`**, at the
/// other end of the Job: that is the gate before anything runs, and this is the
/// decision after it has.
///
/// 409 anywhere but `awaiting_review`, which is what keeps it from becoming the
/// dispatch gate under a second name.
pub(crate) async fn approve_review<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().approve_review(job.id()).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The person merges the pull request their Job opened, and Fleet runs what
/// follows. **`approve_review` with a write to somebody else's repository in
/// front of it** — and the write is what makes the after-merge Checks run.
///
/// 409 anywhere but `awaiting_review`, and 409 where the record holds no pull
/// request. A forge that would not merge answers with the kind it refused on,
/// each its own code: a protected base, a conflict and a required check are
/// 409s, and a missing tool or an unreadable refusal are 500s. Nothing retries.
pub(crate) async fn merge_pull_request<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().merge_pull_request(job.id()).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Send the work back with a note. **The Job comes back `running`**, at the
/// same step, with the same Drone — nothing was thrown away and nothing was
/// spawned.
///
/// 409 where the Drone is gone: there is nobody to tell.
pub(crate) async fn request_changes<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    let note: ChangesRequested = match ipc::decode("a review note", &body) {
        Ok(note) => note,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().request_changes(job.id(), note).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The comments a person picked off the pull request reach a Drone.
///
/// **The same answer `request_changes` gives**, because that is what it calls:
/// the Job comes back `queued` with the words waiting for the Drone
/// re-admission puts on it.
///
/// 409 where a comment named is one a Drone on this Job has already been
/// handed, or one the pull request no longer has. 422 where the press named no
/// comment at all.
pub(crate) async fn take_up_remarks<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    let picked: RemarksTakenUp = match ipc::decode("the comments picked", &body) {
        Ok(picked) => picked,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().take_up_remarks(job.id(), picked).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// A verdict on the work, and the Job is over. **Terminal**, which is what
/// separates it from `request_changes` — and it is not `kill_job`, which clears
/// the Board and carries no verdict at all.
pub(crate) async fn reject_job<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().reject_job(job.id()).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The Judge refused, a person disagrees, and the step advances anyway. **The
/// Job comes back `running`** at the step that follows, with everything the
/// refused Drone did still on the branch.
///
/// 409 anywhere but an `escalated` Job stopped on `gate_failure`: a gate that
/// never weighed the work and a gaming flag are not opinions to be overruled,
/// and a failed mechanical Check is terminal and reaches this route as a Job
/// with no stopped step. 422 on a blank reason.
pub(crate) async fn override_verdict<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    let overruling: Overruled = match ipc::decode("an override", &body) {
        Ok(overruling) => overruling,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().override_verdict(job.id(), overruling).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The gate could not decide, and a person asks it again on the evidence the
/// step already submitted. **The Job comes back wherever the second reading
/// left it** — running at the next step where it ruled, escalated again where
/// it could not.
///
/// No body, because nothing is being disagreed with. 409 anywhere but an
/// `escalated` Job stopped on `gate_undecided`, and 409 again on a Job this
/// Fleet is no longer standing at, where the baseline the first reading used
/// went with the slot and `restart_step` is what applies.
pub(crate) async fn rerun_gate<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().rerun_gate(job.id()).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Ask a Job to show its work: the repository's harness runs the last spec a
/// Drone named, in the Job's worktree, and what it captured comes back as a set
/// of its own. **The Job does not move**, so what comes back is the set and not
/// the Job.
///
/// 409 before anything runs where it cannot — no harness, no worktree, no spec,
/// the spec gone, a Drone working, a press already out. 200 with `nothing` set
/// where the harness ran and captured nothing, which is an answer rather than a
/// refusal. **It may take as long as the app takes to start**: the daemon runs
/// it off the turn loop, and the run finishes whether or not this request is
/// still waiting.
pub(crate) async fn show_again<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.shared().show_again(job.id()).await {
        Ok(shown) => answer(StatusCode::OK, &shown, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Give this Job a higher cost ceiling than the tier above it allows. **The
/// Job comes back exactly where it was** — the raise moves a number and nothing
/// else, and the field a caller reads it for is `queued_reason`: `over_budget`
/// before, absent or `waiting_on_resources` after.
///
/// 409 on a terminal Job, where nothing is left to spend. 422 on a figure at or
/// under the cap in force — the act only ever raises — and 422 again where the
/// surface asked past its own ceiling, which a person never meets.
pub(crate) async fn raise_cost_cap<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    let raise: CapRaise = match ipc::decode("a cost cap", &body) {
        Ok(raise) => raise,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().raise_cost_cap(job.id(), raise).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The other ceiling, and the same three refusals in the same order. 409 on a
/// terminal Job, 422 on a figure at or under the cap in force, 422 again where
/// the surface asked past its own ceiling.
///
/// **A route of its own and never a field on the one above.** The two ceilings
/// are refused separately and cleared separately, so a caller that raised one
/// and was told about the other could not act on the answer.
pub(crate) async fn raise_turn_cap<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    let raise: TurnRaise = match ipc::decode("a turn cap", &body) {
        Ok(raise) => raise,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().raise_turn_cap(job.id(), raise).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The process, not the unit of work. What comes back is the Job the Drone was
/// on, which is still there.
pub(crate) async fn kill_drone<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().kill_drone(job.id()).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// The unit of work, not the process. A separate operation from `kill_drone`
/// because two of the edges into `killed` leave a status no
/// Drone has ever existed under, and neither one can be spelled as killing a
/// Drone.
pub(crate) async fn kill_job<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().kill_job(job.id()).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Go and look at this Job now, and answer with what was found.
///
/// **200 on every answer, including `not_working`.** The examination ran and
/// this is what it found; a status code saying the request failed would make a
/// finding indistinguishable from Fleet being unable to look.
///
/// **A POST, and an act rather than a read.** It writes what it found into the
/// Job's own log, so the answer is on the record beside everything else Fleet
/// did — and it does real work, which is not what a GET promises.
pub(crate) async fn examine_job<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().examine_job(job.id()).await {
        Ok(examined) => answer(StatusCode::OK, &examined, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Delete the Job's whole record. **Real deletion, not a status** — the row
/// and everything beneath it are gone, and there is no undo.
///
/// 409 where the Job is not yet terminal: `kill_job` is the act that ends one
/// still in flight, and this only clears a Board of work that has already
/// finished. It does not touch the worktree or the branch — `armada clean`
/// keeps that concern.
pub(crate) async fn forget_job<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().forget_job(job.id()).await {
        Ok(forgotten) => answer(StatusCode::OK, &forgotten, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Give the Job's worktree and branch back. **The record survives** — this is
/// disk, not the row, and `forget_job` is the act that clears the Board.
///
/// 409 where the Job is not yet terminal, for `forget_job`'s reason. 200 where
/// either half was left standing: a branch holding commits the base cannot
/// reach is kept on purpose, and the answer names it rather than failing.
pub(crate) async fn reclaim_worktree<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().reclaim_worktree(job.id()).await {
        Ok(reclaimed) => answer(StatusCode::OK, &reclaimed, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Mint a replacement for a stopped Job. **Two Jobs come back**, and
/// the answer is 200 rather than 201 because the act a caller asked for is the
/// recovery, not the creation — the new Job's id is in the body.
pub(crate) async fn redispatch_job<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
) -> Response {
    match served.daemon().redispatch_job(job.id()).await {
        Ok(both) => answer(StatusCode::OK, &both, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Say something to the Drone that is there. **The Job comes back `running`**,
/// at the same step, with the same Drone — nothing was spawned and nothing was
/// thrown away.
///
/// 409 where the Drone is gone, naming `restart_step` as the act that applies.
pub(crate) async fn redirect_drone<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    let instruction: Redirection = match ipc::decode("a redirect", &body) {
        Ok(instruction) => instruction,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().redirect_drone(job.id(), instruction).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// Answer the question a Drone asked. **The Job comes back unchanged**: what
/// moved is the Drone, handed the answer as a turn. 409 where nothing waits,
/// where the id names an answered question, or where the label was not offered.
pub(crate) async fn answer_question<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    let chosen: ChosenAnswer = match ipc::decode("an answer", &body) {
        Ok(chosen) => chosen,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    let job_id = job.id();
    match served.daemon().answer_question(job_id, chosen).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}
/// Put a new Drone on the worktree the last one left, and say what to do
/// differently. **One Job comes back**, not two — this is the same Job
/// resuming, which is the whole of what makes it different from a redispatch.
///
/// **The body is optional, and no body is the plain restart.** It is the one
/// route here that reads its bytes conditionally, and the condition is that the
/// act existed without them: a restart with nothing to say sends exactly what
/// it sent before this route learned to read anything, so the commonest use of
/// it did not get a step longer. Bytes that arrive are read as a note and a
/// note that will not parse is the transport's 400 like every other body.
///
/// 409 where the Drone is alive, where the worktree is gone, and where a note
/// is already waiting on the Job. 422 on a note with nothing in it — which is
/// not the same request as one with no note.
pub(crate) async fn restart_step<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    // Emptiness is read here rather than through an `Option<Json<_>>`
    // extractor, because the question this asks is about the bytes and not
    // about their shape: a caller that sends no body and one that sends a
    // zero-length one have both said the same thing, and both predate this.
    let note: Option<RestartRequested> = if body.is_empty() {
        None
    } else {
        match ipc::decode("a restart note", &body) {
            Ok(note) => Some(note),
            Err(why) => return undecodable(&why.to_string(), served.run_id()),
        }
    };
    match served.daemon().restart_step(job.id(), note).await {
        Ok(job) => answer(StatusCode::OK, &job, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

/// A person says this Job failed in error. **201, because a report now
/// exists** — and nothing else does: no Job is proposed, no Drone is spawned,
/// and the Job it names is exactly where it was.
///
/// 422 on a blank sentence. The record was already served by three other
/// routes before anybody pressed anything, so a filing with the bundle and no
/// sentence has added nothing at all.
pub(crate) async fn file_report<D: Commands>(
    State(served): State<Served<D>>,
    job: Resolved,
    body: Bytes,
) -> Response {
    let filing: FileReport = match ipc::decode("a report", &body) {
        Ok(filing) => filing,
        Err(why) => return undecodable(&why.to_string(), served.run_id()),
    };
    match served.daemon().file_report(job.id(), filing).await {
        Ok(report) => answer(StatusCode::CREATED, &report, served.run_id()),
        Err(refusal) => refused(refusal),
    }
}

File: packages/shell/src/failures/commands.ts
// What came back from a command, where what came back was a failure.
//
// **Two builders, and they are the two arms of one decision at the call site**:
// a refusal is Fleet declining with an envelope, and a transport failure is a
// command it did not answer at all. Everything else an `Outcome` carries is the
// form saying what it will not send, which is guidance and takes an `Alert`.
//
// **The only failures a person provoked**, which is why each names what
// survives before it names what to do: a command that may or may not have been
// carried out is the one place pressing again makes two Jobs.
//
// Four codes. Three are minted here, and the fourth is Fleet's own — the one
// code on this surface Bridge did not mint, and does not parse.

import type {
  BridgeCode,
  DebugField,
  FailureDetail,
  FailureMachineValue,
} from "@armada/components";

import type { BridgeIdentity } from "@armada/protocol";
import type { Outcome, WireError } from "@armada/protocol";
import { elapsed } from "../fleet";
import type { Failure } from "./notice";
import { logField, machineLog, versions } from "./notice";

/**
 * Fleet's run id, and only Fleet's.
 *
 * **It is labelled "Fleet run" because that is what it is.** Measured against a
 * live daemon: the id is minted once per Fleet process and every answer clones
 * it, so four different refusals in one session quote one value. Calling it an
 * error id or a reference would promise that it points at this failure, and the
 * first person to quote it would find it names a whole session.
 *
 * Nothing else on this surface carries one. Bridge mints none — an id minted by
 * a process that writes no log line joins to nothing — and a renderer crash
 * never went near Fleet, so it has none to show and shows none.
 */
function fleetRun(runId: string): FailureMachineValue[] {
  if (runId === "") return [];
  return [{ value: runId, copyValue: runId, meta: "Fleet run" }];
}

/**
 * A command Fleet refused.
 *
 * **The only failure here that carries a run id**, because it is the only one
 * minted on the other side of the connection. It names Fleet's run, and the row
 * says so.
 *
 * **The one failure whose code Bridge did not mint**, and nothing here
 * interprets it. It is opaque to Bridge — looked up, never parsed — and the
 * message is what renders when the lookup misses. It carries no `bridge.`
 * prefix, which is how a reader tells at a glance that Fleet raised it and a
 * manifest holds what it means.
 *
 * The whole `fields` map and the whole `chain` are folded away rather than
 * summarised, because a refusal's `message` names one problem even where
 * several exist.
 */
export function refusalFailure(error: WireError, bridge: BridgeIdentity): Failure {
  const details: FailureDetail[] = [
    { label: "Code", value: error.code },
    { label: "Message", value: error.message },
  ];
  if (error.job_id !== undefined) details.push({ label: "Job", value: error.job_id });
  if (error.drone_id !== undefined) details.push({ label: "Drone", value: error.drone_id });
  if (error.step_id !== undefined) details.push({ label: "Step", value: error.step_id });
  for (const [key, value] of Object.entries(error.fields ?? {})) {
    details.push({ label: key, value: String(value) });
  }
  if ((error.chain ?? []).length > 0) {
    details.push({ label: "Chain", value: error.chain.join("\n") });
  }

  return {
    // **A fault, read off the situation and not off the code.** A command was
    // refused, so Armada did not do the thing — and the code here is opaque,
    // so deriving a class from it would read a value the contract forbids.
    kind: "fault",
    // The only one of the six with everything the contract guarantees. The
    // wire's `fields` keys pass through with their own spelling: they are what
    // somebody greps a log for, and rewriting them into prose would break the
    // one join this payload exists to make.
    payload: {
      code: error.code,
      message: error.message,
      run_id: error.run_id,
      ...(error.job_id === undefined ? {} : { job_id: error.job_id }),
      ...(error.drone_id === undefined ? {} : { drone_id: error.drone_id }),
      ...(error.step_id === undefined ? {} : { step_id: error.step_id }),
      fields: [
        ...Object.entries(error.fields ?? {}).map(([key, value]) => ({
          key,
          value: String(value),
        })),
        ...logField(bridge),
      ],
      chain: error.chain ?? [],
      ...versions(bridge),
    },
    headline: error.message,
    next: "Nothing was sent. Change what the command names, or read the log.",
    detailsLabel: "What Fleet refused",
    details,
    values: [...machineLog(bridge), ...fleetRun(error.run_id)],
    note: "The run names Fleet's process for this session, not this one failure. It is what joins this to Fleet's log lines.",
  };
}

/** Bridge sent a command and no answer came back inside the wait. */
const COMMAND_TIMED_OUT: BridgeCode = "bridge.command.timed_out";

/** The request itself failed, so whether Fleet read it is unknown. */
const COMMAND_UNREACHABLE: BridgeCode = "bridge.command.unreachable";

/** Fleet answered a status, and the body under it was not a refusal. */
const COMMAND_UNANSWERABLE: BridgeCode = "bridge.command.unanswerable";

/**
 * A command Fleet did not answer.
 *
 * **The sixth failure, and the one that used to be a line of text.** A refusal
 * carries a `WireError` and reaches `refusalFailure`; everything else on the
 * same seam — a wait that ran out, a socket that failed, a status with no
 * refusal under it — was drawn as a single `Alert` reading "Fleet did not
 * answer: <the machine's words>". That sentence has no code, no fold, and
 * nothing to copy, so a person who hit it had one line and no way to hand it
 * on. It is the generic message this file exists to repair, and it survived
 * here because a transport failure is the one seam failure with no envelope.
 *
 * **Bridge mints the code, and the `bridge.` prefix is what keeps that honest**
 * — the fault may well be Fleet's, and what Bridge is naming is the condition
 * it observed. Three of them, because the three take three different next
 * steps, which `TransportFault` states.
 *
 * **No run id, on any of the three.** Two never got an answer to carry one, and
 * the third got a body Bridge could not read one out of. A labelled blank would
 * claim Bridge looked and Fleet sent none.
 */
export function transportFailure(
  outcome: Extract<Outcome, { ok: false; why: "transport" }>,
  bridge: BridgeIdentity,
): Failure {
  const fault = outcome.fault;
  const asked = `${fault.method} ${fault.path}`;
  const base = {
    // **All three are faults**, whatever Fleet's state turns out to be: what
    // failed is a command rather than a view. The third of them proves Fleet
    // alive and answering and is still one, because the act did not complete
    // and no wait completes it. The degraded thing on this seam is the
    // *connection*, which `fleetFailure` draws and `unreachable` points at.
    kind: "fault" as const,
    detailsLabel: "What Bridge asked",
    values: machineLog(bridge),
  };
  const fields: DebugField[] = [
    { key: "why", value: fault.why },
    { key: "method", value: fault.method },
    { key: "path", value: fault.path },
    { key: "detail", value: outcome.detail },
  ];
  const details: FailureDetail[] = [
    { label: "Route", value: asked },
    { label: "Detail", value: outcome.detail },
  ];

  switch (fault.why) {
    case "timed_out": {
      const waited = elapsed(fault.waitedMs);
      return {
        ...base,
        headline: `Fleet did not answer ${asked} inside ${waited}`,
        // **Never "nothing was sent".** The request reached Fleet and the wait
        // ran out on this side, so the act may have been carried out — and a
        // sentence telling somebody to send it again would be how a request
        // becomes two Jobs. The board is the thing that knows.
        next: "The command may still have been carried out. Bridge will read the board again shortly; read it yourself before sending it a second time.",
        payload: {
          code: COMMAND_TIMED_OUT,
          message: `Fleet did not answer ${asked} inside ${waited}`,
          fields: [...fields, { key: "waited_ms", value: String(fault.waitedMs) }, ...logField(bridge)],
          ...versions(bridge),
        },
        details: [...details, { label: "Waited", value: waited }],
        note: "The wait is Bridge's, not Fleet's. A route with a model call inside it answers within a budget of its own and a refusal from it would carry a code — an ordinary command has no such budget, so a slow answer here is not necessarily a stuck Fleet. There is no Fleet run id to quote either way: this is Bridge giving up first.",
      };
    }

    case "unreachable":
      return {
        ...base,
        headline: `Bridge could not reach Fleet for ${asked}`,
        // Deliberately not "nothing was sent": a socket that failed may have
        // failed after the bytes went out, and deciding otherwise on no
        // evidence is what makes a duplicate.
        next: "Whether Fleet read it is unknown. Read the board, then send it again.",
        payload: {
          code: COMMAND_UNREACHABLE,
          message: outcome.detail,
          fields: [...fields, ...logField(bridge)],
          ...versions(bridge),
        },
        details,
        note: "The connection failed rather than the command. If the status bar also says Fleet is unreachable, that notice is the one worth sending.",
      };

    case "unanswerable":
      return {
        ...base,
        headline: `Fleet answered ${fault.status} on ${asked}, and Bridge could not read it`,
        // Fleet is up. Retrying sends the same request down the same route to
        // the same disagreement, so the sentence points at the log instead.
        next: "Fleet is running and answered. Nothing here will change by sending it again — read the log.",
        payload: {
          code: COMMAND_UNANSWERABLE,
          message: `Fleet answered ${fault.status} with a body that is not a refusal`,
          fields: [...fields, { key: "status", value: String(fault.status) }, ...logField(bridge)],
          ...versions(bridge),
        },
        details: [...details, { label: "Status", value: String(fault.status) }],
        // The two protocol versions are already on the payload's tail, and
        // this is the one failure where they are the first thing to look at.
        note: "A status with no refusal under it is the two sides disagreeing about the route, not a job going wrong. The protocol versions on this record are what to check first.",
      };
  }
}
